Build: standardize output directory for artifacts

humitos opened the pull request:
Always expose the final artifacts under `_readthedocs/<format>` directory, where `<format>` can be one of the following:

- html
- htmlzip
- pdf
- epub
- json

This allows all users of the platform to have the same behavior. No matter if they are using the default Sphinx/MkDocs builders or a custom one by using `build.commands`.

Besides, this commit removes the "intermediary" directory used _before_ uploading the files to S3. This allows users to modify the output files directly from `_readthedocs/<format>` using `build.jobs.post_build`, for example.

### Some important notes

- `BaseBuilder.type` is removed since it was not required anymore.
- There is no `old_artifact_path` anymore since there is no intermediary folder.
- The `sphinx-build` command is executed from path where the repository is cloned, instead from the the directory where the `conf.py` lives. This is to simplify the code and be able to use relative path for the build output directory (e.g. `sphinx-build -b html ... docs/ _readthedocs/html`).
- A new `BaseSphinx._post_build` method was added to cleanup the output directory (leave only one .epub, .html and .pdf file before moving forward).
- These changes should allow us to support _all formats_ for `build.commands`. We would need to check if these standard directories exist, set these formats in the version and upload their content.
- Upload step now validate there is one and only one PDF, ePUB and ZIP file when they are built. Currently, we only support one file per version. In the future we could remove this restriction. It fails silently for now, which is not great. We need to refactor this code a little to fail the build properly and communicate this to the user. I think it's _the correct way_ of doing it.
- This could also allow to support other ways to build PDF (SimplePDF or `tectonic`, see https://github.com/readthedocs/readthedocs.org/issues/9598) and also ZIP (e.g. zundler, see https://github.com/readthedocs/meta/issues/77#issuecomment-1301927685).

### Use cases manually tested from `test-builds`

* `mkdocs-python-tags`
* `build-commands`
* `all-formats`
* `latest`
* `build-jobs-post-build`
* `singlehtml`
* `htmldir`

Closes https://github.com/readthedocs/readthedocs.org/issues/9179

humitos commented on readthedocs/doc_builder/backends/sphinx.py:
```diff
@@ -390,65 +376,62 @@ def __init__(self, *args, **kwargs):
                 )
         archive.close()
 
+        safe_rmtree(sphinx_build_dir)
```
This produces tests to fail with this exception:

```
  File "/usr/src/app/checkouts/readthedocs.org/readthedocs/doc_builder/backends/sphinx.py", line 407, in _post_build
    safe_rmtree(sphinx_build_dir)
  File "/usr/src/app/checkouts/readthedocs.org/readthedocs/core/utils/filesystem.py", line 146, in safe_rmtree
    _assert_path_is_inside_docroot(path)
  File "/usr/src/app/checkouts/readthedocs.org/readthedocs/core/utils/filesystem.py", line 32, in _assert_path_is_inside_docroot
    raise SuspiciousFileOperation(path)
django.core.exceptions.SuspiciousFileOperation: /tmp/readthedocs-tests/git-repository/_readthedocs/htmlzip
```

It's totally fine to delete that directory. @stsewd why we are getting this error here? I think this line should be inside the `if path.is_symlink()`, right? https://github.com/readthedocs/readthedocs.org/blob/main/readthedocs/core/utils/filesystem.py#L146

stsewd replied:
During tests, you need to override the DOCROOT setting to point to your temp directory.

https://github.com/readthedocs/readthedocs.org/blob/84a149a57ccd9c1b5cd47bb285263a273124a199/readthedocs/config/tests/test_config.py#L696-L696

ericholscher commented:
> * The `sphinx-build` command is executed from path where the repository is cloned, instead from the the directory where the `conf.py` lives. This is to simplify the code and be able to use relative path for the build output directory (e.g. `sphinx-build -b html ... docs/ _readthedocs/html`).

This one scares me a little bit, as I mentioned in the call. Users are doing lots of weird things with the working directory in `conf.py`, mostly to get various things added to the PYTHONPATH or similar. I do think we probably want to avoid this if we can easily undo it.

ericholscher commented on readthedocs/doc_builder/backends/sphinx.py:
```diff
@@ -37,6 +37,8 @@ class BaseSphinx(BaseBuilder):
 
     """The parent for most sphinx builders."""
 
+    sphinx_doctrees_dir = "_build/doctrees"
```
This needs a comment I think explaining what it's for. Also seems like it should probably not be plural?

humitos replied:
I added a small comment based on the Sphinx's documentation. The plural form comes from the default value of this setting `.doctrees`. Note that I haven't changed the name of this directory in this PR, I just moved it to a class variable instead of hardcode it in multiple places.

ericholscher commented on readthedocs/doc_builder/backends/sphinx.py:
```diff
@@ -48,16 +50,8 @@ def __init__(self, *args, **kwargs):
-                docs_dir,
-                self.sphinx_build_dir,
-            )
+            pass
```
This definitely needs a comment.

humitos replied:
I may need to research a little more here to be able to add a comment 😄 . I deleted this code because `docs_dir` nor `self.old_artifact_path` were used anymore so I had to add a `pass`. However, I'm not sure what the try/except block is catching --maybe it can be removed as well? 🤔

humitos replied:
I added a big note explaining what's happening here. Note that _I'm not changing the current behavior_. However, I do not agree with the current behavior and I think we should refactor/change it to hard fail the build in of the cases and the other should probably managed differently (in general, I'd refactor the whole `Project.config_file` method and how it's used from outside).

All of this is outside the scope of this PR, tho, so I'm not making these changes here and I hope that the note I added to be enough for now.

ericholscher commented on readthedocs/doc_builder/backends/sphinx.py:
```diff
@@ -264,7 +258,6 @@ def append_conf(self):
         )
 
     def build(self):
-        self.clean()
```
I know this was probably not needed, but I'd like to reduce moving pieces here and not try to "clean up" too much and randomly break things when we're already doing a large refactor.

humitos replied:
I removed it because there is no "old artifact path" (variable `self.old_artifact_path`) to remove anymore. This is because we only have one output directory now (`_readthedocs/<format>`). Note the super `BaseBuilder.clean()` method was completely removed.

ericholscher commented on readthedocs/doc_builder/backends/sphinx.py:
```diff
@@ -273,22 +266,34 @@ def build(self):
+            build_command.extend(["-W", "--keep-going"])
+        build_command.extend(
+            [
+                "-b",
```
We should use long-form commands for all these, so it's clearer in the output I think?

ericholscher replied:
Ugh, of course.. 🙃

ericholscher commented on readthedocs/doc_builder/backends/sphinx.py:
```diff
@@ -273,22 +266,34 @@ def build(self):
+                f"language={project.language}",
+                # Execute the command at the clone directory,
+                # but use `docs/` instead of `.` for the Sphinx path
+                os.path.dirname(
+                    os.path.relpath(
+                        self.config_file,
+                        self.project_path,
+                    ),
+                ),
+                self.sphinx_build_dir,
```
These could both use comments explaining what the argument is, I think?

humitos replied:
Added a comment saying they are the SOURCEDIR and OUTPUTDIR 👍🏼

ericholscher commented on readthedocs/doc_builder/backends/sphinx.py:
```diff
@@ -273,22 +266,34 @@ def build(self):
-            cwd=os.path.dirname(self.config_file),
             bin_path=self.python_env.venv_bin(),
+            cwd=self.project_path,
         )
```
Seems like this is where we can remove this change in `cwd`? I'm 👍 on undoing this change.

humitos replied:
I revert this change to use the CWD we were using previously.

ericholscher commented on readthedocs/doc_builder/backends/sphinx.py:
```diff
@@ -337,40 +342,19 @@ def venv_sphinx_supports_latexmk(self):
         if self.project.has_feature(Feature.USE_SPHINX_BUILDERS):
             self.sphinx_builder = 'html'
 
-    def move(self, **__):
-        super().move()
```
🔥

ericholscher commented on readthedocs/doc_builder/backends/sphinx.py:
```diff
@@ -390,64 +373,68 @@ def __init__(self, *args, **kwargs):
+            cwd=self.project_path,
+            record=False,
+        )
+        self.run("rm", "-r", self.sphinx_build_dir, cwd=self.project_path, record=False)
```
This scares me a little bit. Could users have already written something here and expected it to stick around? 🤔

humitos replied:
Ideally, if people wants to add/modify anything to/from the output directory they should use `post_build`. We should probably clarify the output directory will be cleaned out _before_ immediately before the `build` step. What do you think?

ericholscher commented on readthedocs/doc_builder/backends/sphinx.py:
```diff
@@ -390,64 +373,68 @@ def __init__(self, *args, **kwargs):
+
+    def _post_build(self):
+        """Internal post build to cleanup EPUB output directory and leave only one .epub file."""
+        sphinx_build_dir = os.path.join(self.project_path, self.sphinx_build_dir)
```
Seems odd to overwrite this class-level variable with  something the same name. Should the class-level one be more explicit that it's relative? 

```suggestion
        sphinx_build_dir = os.path.join(self.project_path, self.relative_build_dir)
```

Also seems odd it's named `sphinx` when nothing about it is sphinx specific?

humitos replied:
I'm using `relative_output_dir` and `absolute_output_dir` now. I think it's more explicit and will avoid confusions.

ericholscher commented on readthedocs/doc_builder/backends/sphinx.py:
```diff
@@ -390,64 +373,68 @@ def __init__(self, *args, **kwargs):
+        epub_sphinx_filepaths = glob(os.path.join(sphinx_build_dir, "*.epub"))
+        if epub_sphinx_filepaths:
+            # NOTE: we currently support only one .epub per version
+            epub_filepath = epub_sphinx_filepaths[0]
```
This seems like another place to raise a warning/error if multiple are generated?

humitos replied:
Yes and... no 😅 . I decided to not raise an exception here nor at the upload step either because that will change the current behavior. That was basically the main reason to not fail the builds because of this.

This code is mostly copied/adapted from the previous `move()` method that was not failing the build: https://github.com/readthedocs/readthedocs.org/blob/5252be35fb12c4222f56db356af6e92be2fd04c6/readthedocs/doc_builder/backends/sphinx.py#L437-L445

humitos replied:
I added a warning log to collect what are the projects generating multiple ePubs and be able to communicate to them in the future in case we want to; for example to tell them "We now support multiple ePubs" 👍🏼

ericholscher replied:
Cool 👍

ericholscher commented on readthedocs/doc_builder/backends/sphinx.py:
```diff
@@ -479,30 +466,36 @@ class PdfBuilder(BaseSphinx):
             *self.get_sphinx_cmd(),
-            '-b',
-            'latex',
+            # New arguments for standardization
```
This comment could use work. It doesn't tell me what it actually does or why. Standardization with what?  Seems like it's enabling full tracebacks? https://www.sphinx-doc.org/en/master/man/sphinx-build.html#cmdoption-sphinx-build-T

humitos replied:
He he, this was more an internal to comment in the PR description, that I finally forget removing and commenting in the description 😓 . I added this arguments to match the ones we are using in all the other Sphinx commands. I'm happy to rever this, but I think it's correct to standardize them on all the commands. We have added some arguments to the other commands over time but we haven't add those in this one.

humitos replied:
Let me know if it's fine keeping them or I should remove them for now.

ericholscher commented on readthedocs/doc_builder/backends/sphinx.py:
```diff
@@ -632,40 +628,41 @@ def _build_pdflatex(self, tex_files, latex_cwd):
-            self.old_artifact_path,
-            '{}.pdf'.format(self.project.slug.capitalize()),
+        if not self.pdf_file_name:
+            log.info("PDF file was not generated/found.")
```
This seems odd again to not raise this to the users attention?

humitos replied:
Yeap. You are correct here. I added the code to raise an exception in this case and also to log a warning if there are multiple PDF generated (following the same idea than for the ePubs)

ericholscher commented on readthedocs/doc_builder/base.py:
```diff
@@ -60,27 +49,8 @@ def build(self):
-            safe_rmtree(self.old_artifact_path)
-            log.info('Removing old artifact path.', path=self.old_artifact_path)
+    def _post_build(self):
+        """Execute extra steps (e.g. create ZIP, rename PDF, etc) after building if required."""
```
Does Python support empty function bodies?! 

I generally still think it's cleaner to have a `pass` or `raise NotImplementedError` here to be more explicit.

humitos replied:
He, yes. It does. I usually add a `pass` as well. However, it seems that Black removed it. We cannot have a `raise NotImplementedError` because it's always called by the builder and some of the subclass don't implement it.

I'm not sure if there is a way to tell Black that I want to keep the `pass` in this line. That's probably what we need here.

ericholscher commented on readthedocs/projects/tasks/builds.py:
```diff
@@ -514,28 +512,38 @@ def on_failure(self, exc, task_id, args, kwargs, einfo):
         )
 
-        # NOTE: we are updating the db version instance *only* when
         # HTML are built successfully.
```
Seems we deleted half this comment. Is it not true anymore? 

Ah. I see it got missed in a comment refactor. This line needs to be moved down below :)

humitos replied:
I got confused by this comment. I'm not sure if there is something I have to do here or not 😅

ericholscher replied:
Yes, the comment was 2 lines, and only 1 line was moved a few lines down. The comment is basically saying we only hit the API when `html` passes, which is medium-weird, but we don't want to change with this refactor.

humitos replied:
> which is medium-weird, but we don't want to change with this refactor

Yes, this is _wrong_, in my opinion. I noticed this when I did the Celery refactor. This is something I want to change soon, if possible.

ericholscher commented on readthedocs/projects/tasks/builds.py:
```diff
@@ -798,34 +793,44 @@ def store_build_artifacts(
+            ):
+                types_to_copy.append(artifact_type)
+            # Never delete HTML nor JSON (search index)
+            elif artifact_type not in ("html", "json"):
```
I think it's probably clearer to have a constant like `UNDELETABLE_TYPES` that we're checking here. This seems like something that might get missed in a future update, and we should have a bit more comment on why we never delete these.

humitos replied:
_Good_ point. I added constants for this.

ericholscher commented on readthedocs/projects/tests/test_build_tasks.py:
```diff
@@ -197,6 +197,11 @@ def test_build_updates_documentation_type(self, load_yaml_config):
             }
         )
 
+        # Create the artifact paths, so it's detected by the builder
```
Could use a bit more context here. I think you're trying to say:

```suggestion
        # Create the artifact paths, so that our mocked objects
        # get picked up as having been "built".
```

humitos replied:
Hrm, it's not for the mocked objects. We are doing checks for path existence to know if a format was built in the build or not to decide if it has to be uploaded to the storage. This is done at https://github.com/readthedocs/readthedocs.org/blob/faa611fad689675f81101ea643770a6b669bf529/readthedocs/projects/tasks/builds.py#L798-L804

I'm not sure how to make it clearer. I'd appreciate if you have a suggestion to expand that comment now you have a little more context about the logic behind it.

ericholscher replied:
Updated the comment.

ericholscher reviewed (commented):
I'm very excited about this change, and think it's a great one. Other than my 2 comments above, I just had a few small nitpicks as I read through things. Feel free to disregard some of them, I am just a bit hesitent to change too much all at once, given the side of this already underlying change.

Really think this is a great step forward overall though. We should coordinate this change w/ the rclone changes @stsewd is making. I don't think they'll conflict too badly, but definitely changing a lot of similar code.

stsewd commented:
> We should coordinate this change w/ the rclone changes @stsewd is making. I don't think they'll conflict too badly, but definitely changing a lot of similar code.

This doesn't conflict with the rclone PR.

humitos commented:
@ericholscher 

> This one scares me a little bit, as I mentioned in the call. Users are doing lots of weird things with the working directory in `conf.py`, mostly to get various things added to the PYTHONPATH or similar. I do think we probably want to avoid this if we can easily undo it.

This is definitely a good point.

I can revert this change, but it will produce different relative output paths depending on the location of the `conf.py` for the project --which could be confusing:

* a project with `docs/conf.py` will execute this command: `sphinx-build -b html [...] . ../_readthedocs/html`
* another project with `src/docs/user/conf.py` will call: `sphinx-build -b html [...] . ../../../_readthedocs/html`

Note we will be keeping the `source_directory` as `.` but they will have different `output_directory`. I decided to go the way I wrote the code because it's more explicit, easier to read and the command it's consistent across all the projects (instead of changing the `output_directory`, this implementation changes the `source_directory` which makes it explicit, since the `cwd` is not exposed to users)

_In theory_, people should be modifying the `PYTHONPATH` based on the location of the `conf.py` file (since their code is running from inside it) instead of `cwd`; but I know that's not true _in practice_ 😄

humitos commented:
@ericholscher 

> I think the next step here would probably be supporting a `zip` file, to allow people to upload multiple? Or support multiple, and zip them on our side. But not a problem for this PR :)

I prefer supporting uploading _and serving_ multiple `.pdf` as-is. Otherwise, people will have to download a big `.zip` just to read one small PDF. Being able to read _just one_ PDF is a good feature. Also, giving people the ability to "download all the PDF as .zip" is another good feature.

In summary, I want both of your suggestions 😄

humitos commented:
@ericholscher I addressed most of the feedback (thanks!). I reverted the CWD change and I added an exception to raise when there is no PDF found. I repeated some QA locally and it seems that things keep working fine.

Please, let me know if there are more changes to be made on this PR. If there are things we want to address later, we can create issues otherwise.

ericholscher commented:
> I can revert this change, but it will produce different relative output paths depending on the location of the `conf.py` for the project --which could be confusing:
> 
> * a project with `docs/conf.py` will execute this command: `sphinx-build -b html [...] . ../_readthedocs/html`
> * another project with `src/docs/user/conf.py` will call: `sphinx-build -b html [...] . ../../../_readthedocs/html`

I wonder about using a full path for the `readthedocs/html` directory? I agree that it's not a great UX, but probably nicer output to just have it be explicit. Even better if we can just do something like `$HOME/rtd-builds/<project>/<version>/_readthedocs/html`? 🤷 

I agree in the future the more explicit path is nicer, but this is a good baby step to getting there.

ericholscher commented on readthedocs/builds/constants.py:
```diff
@@ -148,3 +148,19 @@
+    "pdf",
+    "epub",
+)
+# Artifacts that are not deleted when uploading to the storage,
+# even if they weren't re-built in the build process.
```
```suggestion
# Artifacts that are not deleted when uploading to storage.
# This is because we want to only upload changed files,
# whereas single-file types will need to be deleted & recreated on any change.
```

I think this is the thinking here. Will this be true once we support multiple PDF's? 🤔

humitos replied:
Hrm, thinking a little more here, I think that both comments (mine and yours) are wrong 😅 

Taking into account that HTML _is mandatory_ for a build, we will always update the HTML output in our storage. Since JSON _is required_ for search we also always want to have it there.

The other formats may be required to delete them when they are _disabled after being enabled_. If the build for version `v3` outputs a PDF and then the user disable the PDF for that version, we have to remove the PDF from the storage. This can't happen for HTML nor JSON because those _are mandatory_.

I complicated myself trying to explain this, but hopefully, I'm being clear enough for you to understand me 😄

ericholscher replied:
I'm not 100% following... But it's probably not too important as long as we have a reasonably correct comment.

ericholscher commented on readthedocs/doc_builder/backends/sphinx.py:
```diff
@@ -40,9 +40,16 @@ class BaseSphinx(BaseBuilder):
 
+    # Output directory relative to where the repository was cloned
+    # (e.g. "_readthedocs/<format>")
+    relative_output_dir = ""
```
Empty string feels like an odd default. I think we probably want `None` to make it blow up when not set properly?

humitos replied:
Yeap. I think I pushed this by mistake while testing why one of the tests was failing. Thanks for pointing it!

ericholscher commented on readthedocs/doc_builder/backends/sphinx.py:
```diff
@@ -496,30 +506,28 @@ def build(self):
-        latex_cwd = os.path.join(self.project_path, self.sphinx_build_dir)
-        tex_files = glob(os.path.join(latex_cwd, '*.tex'))
 
+        tex_files = glob(os.path.join(self.absolute_output_dir, '*.tex'))
```
Nice a clean 👍

ericholscher commented on readthedocs/doc_builder/backends/sphinx.py:
```diff
@@ -496,30 +506,28 @@ def build(self):
         images = []
-        for extension in ('png', 'gif', 'jpg', 'jpeg'):
-            images.extend(latex_path.glob(f'*.{extension}'))
+        for extension in ("png", "gif", "jpg", "jpeg"):
+            images.extend(Path(self.absolute_output_dir).glob(f"*.{extension}"))
```
🙈

ericholscher commented on readthedocs/doc_builder/backends/sphinx.py:
```diff
@@ -275,21 +285,22 @@ def build(self):
-                        self.config_file,
-                        self.project_path,
-                    ),
+                # Sphinx's source directory (SOURCEDIR).
+                # We are executing this command at the location of the `conf.py` file (CWD).
+                # TODO: ideally we should execute it from where the repository was clonned,
+                # but that could lead unexpected behavior to some users:
+                # https://github.com/readthedocs/readthedocs.org/pull/9888#issuecomment-1384649346
+                ".",
+                # Sphinx's output build directory (OUTPUTDIR)
```
Great comments 👍

ericholscher commented on readthedocs/doc_builder/backends/sphinx.py:
```diff
@@ -374,48 +382,65 @@ def __init__(self, *args, **kwargs):
         )
 
-        # FIXME: ``arcname`` is invalid with this command still, but it's close
+        # Move the directory into a temporal directory,
```
This is the security PR fix just getting pulled in, right?

ericholscher replied:
```suggestion
        # **SECURITY CRITICAL: Advisory GHSA-hqwg-gjqw-h5wg**
        # Move the directory into a temporal directory,
```

humitos replied:
Yes.

ericholscher reviewed (commented):
These changes are great, thanks for the full explanations here. I think this is getting close, I have a lot fewer comments this time :)

humitos commented:
@ericholscher 

> I wonder about using a full path for the `readthedocs/html` directory? I agree that it's not a great UX, but probably nicer output to just have it be explicit. Even better if we can just do something like `$HOME/rtd-builds/<project>/<version>/_readthedocs/html` or even `$BUILD_ROOT/_readthedocs/html`??

I really like this idea of `$BUILD_ROOT` 💯 . I'd probably name it differently and also include the `_readthedocs/` part on it. This is my proposal:

```
$READTHEDOCS_OUTPUT=<path where the repository was clonned>/_readthedocs/
```

Then, the Sphinx command executed would be:

```
python -m sphinx -T -E -j auto -b singlehtml -d _build/doctrees -D language=en . $READTHEDOCS_OUTPUT/html
```

I think it's pretty clear, short and easy to understand 👍🏼 . It also doesn't change the contract we have with `build.commands` since it keeps using the `_readthedocs/html` folder from inside where the repository was cloned. Ideally, we should use a path completely separated from the "user's repository" to avoid potential collisions, but that will break the contract we currently have.

ericholscher commented on readthedocs/doc_builder/backends/sphinx.py:
```diff
@@ -61,6 +61,18 @@ def __init__(self, *args, **kwargs):
+            #  - the repository has multiple config files (none of them with `doc` in its filename)
+            #  - there is no config file at all
+            #
+            # IMO, if there are multiple config files,
+            # the build should fail immediately communicating this to the user.
+            # This can be achived by unhandle the exception here
+            # and leaving `on_failure` Celery handle to deal with it.
+            #
+            # In case there is no config file, we should continue the build
+            # because Read the Docs will automatically create one for it.
```
This seems reasonable. Is there a good way to implement the suggested logic? I guess raise a different exception for multiple configs that we `raise` (or just don't catch)?

humitos replied:
Yes. I documented this here in this way because I want to change the internal behavior and also communicate these errors better to the user. I'm opening a new issue with all these findings to continue working after this, since it will require some refactor and more design decisions.

ericholscher reviewed (approved):
I think this looks pretty much ready. It's definitely something we should be testing heavily during QA steps, and ahead of deploy if we have the _energy_. Really excited to see it go out!

humitos commented:
@ericholscher thanks for all these reviews 👍🏼 . I opened https://github.com/readthedocs/readthedocs.org/issues/9918 with all the notes we took from the conversations on this PR to not forget them and decide which of them we want to prioritize. I tried to link as many thing there as I could.

I'm going to merge this PR now and ping the rest of the team so they can do some QA locally before deploying it next week.
